### PR TITLE
#228 - [Participant App] Extend addons validateable input to support async validation (RUI Addon)

### PR DESCRIPTION
**Pull request by kscherling**
### What does this PR do?

This PR extends the validateable input to include an async flag.  If false it will not display error states until the `didValidate` flag has been toggled.  This property must be managed on the controller level.

Additionally it cleans up some of the logic which was making some incorrect assumptions and was hard to reason about.

### How should this be tested?

This will need to be tested on admin and the new participant app.

**Set Up UI**
- Pull latest and `npm link` (make sure you are using `node 4.4.4`)

**Set up discuss**
- Pull `revelation-frontend-discuss` 
- `nvm use 4.4.4`
- `npm link ember-cli-revelation-ui  && npm install`
- `ember s`
- Navigate to `localhost:3000/cli/ea/:project_id/setup/settings/digest/customize

**Set up participant**
- pull `revelation-frontend-participant`
- `nvm use 4.4.4`
- `npm link ember-cli-revelation-ui  && npm install`
- `ember s --environment mobileDev`
- Navigate to `localhost:4202/#/login`

### QA Checklist

**Digest (synchronous validation)**

- [x] Digest inputs show error when empty
- [x] Show valid when not empty and changed
- [x] Validation clears after save

**Participant (asynchronousvalidation)**

- [x] Login fields do not display validation when changed
- [x] When login button clicked, validation shows
- [x] Once validation is triggered, validators will validate asyncronously

### Impacted Areas

- Discuss (Just the settings form and only plain text input fields)
- Participant Login

### Related Issues

Closes #228
https://invent.focusvision.com/Portland/revelation-frontend-participant/issues/228

- [x] QA Complete

### CODE REVIEW

**Comment by kscherling** [comment]
Issue https://invent.focusvision.com/Portland/revelation-frontend-participant/issues/228

File: addon/pods/components/rui-validatable-input/component.js [hunk 1]
```diff
@@ -8,35 +8,103 @@ const {
...
-  })
 
+  // This method ensures that if presence is required
+  // But the initial value is undefined or an empty string
+  // (the case with new records) we can still validate even though
+  // the `_didChange` attr would report it had not
+
+  _validatePresenceWithEmptyDefault: computed('value', function() {
+    const validationOptions = Object.keys(
+      this.get(`model.validations.attrs.${this.get('valuePath')}.options`))
```
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: Can use the `this.get('validation.options')`, we already bound this prop on init.

File: addon/pods/components/rui-validatable-input/component.js [hunk 2]
```diff
@@ -8,35 +8,103 @@ const {
...
+
+  // Base values that compute validation state
+
+  _isInvalid: computed.and('validation.isInvalid', '_validatable'),
+  _isValid: computed.and('validation.isValid', '_validatable'),
+
+  // Checks property `_didValidate` on controller
+  // Used to only show validation if `async` is true
+
+  _didValidate: computed.oneWay('targetObject.didValidate')
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: I feel like this should be a property on the model and not the targetObject.

File: addon/pods/components/rui-validatable-input/component.js [hunk 3]
```diff
@@ -8,35 +8,103 @@ const {
...
+      (this.get('value') === '')
+  }),
+
+  // The attribute has changed UNLESS
+  // the validator includes presence and
+  // the value is undefined or empty
+
+  _didChange: computed(
+    'value',
+    'model.hasDirtyAttributes',
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: this seems unnecessary, since changing the value will already trigger this.

File: addon/pods/components/rui-validatable-input/component.js [hunk 4]
```diff
@@ -8,35 +8,103 @@ const {
...
-  showMessage: computed('validation.isDirty', 'isInvalid', 'didValidate', function() {
-    return (this.get('validation.isDirty') || this.get('didValidate')) && this.get('isInvalid')
-  })
 
+  // This method ensures that if presence is required
+  // But the initial value is undefined or an empty string
+  // (the case with new records) we can still validate even though
+  // the `_didChange` attr would report it had not
+
+  _validatePresenceWithEmptyDefault: computed('value', function() {
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: the naming of this confuses me a lot...I kinda know and am still struggling to get my head around it.

File: addon/pods/components/rui-validatable-input/template.hbs [hunk 5]
```diff
@@ -11,9 +11,8 @@ name=valuePath}}
   </p>
 {{/if}}
 
-{{#if showMessage}}
+{{#if isInvalid}}
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/template.hbs]: I feel like there should be an option to hide the messages, but I don't know if that is part of the scope of this.

**Comment by kylemellander** [comment]
@kscherling done commenting - just a few things that I think could clean it up a bit, but overall, pretty solid.

File: addon/pods/components/rui-validatable-input/component.js (hunk 3 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: I wondered that too.  Some of these are holdovers from the previous implementation but I was hesitant to remove.  I will do a bit more testing...

File: addon/pods/components/rui-validatable-input/component.js (hunk 2 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: There is a good bit of discussion on this topic.  Either way we will likely need to manually set it on the model or the targetObject.  I did a bit of research into hooking to store events but this still won't solve the non-model based validations.  But I agree that it would likely be a little less brittle?

I wish this was set when .validate() is called.

File: addon/pods/components/rui-validatable-input/component.js (hunk 4 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: Ha, and funny that I thought that I was being overly explicit!

File: addon/pods/components/rui-validatable-input/template.hbs (hunk 5 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/template.hbs]: I am still working on the styling for this and will be included in a separate pr.

File: addon/pods/components/rui-validatable-input/component.js (hunk 2 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: leaving this for now.

File: addon/pods/components/rui-validatable-input/component.js (hunk 3 above)
Comment (kscherling) [inline addon/pods/components/rui-validatable-input/component.js]: Ah, so this is needed for when you save a model.  If not the validations will not clear after the model is persisted flag flips.

Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: ahhhh 👍

**Comment by kylemellander** [comment]
+1

**Comment by alexbeeken** [comment]
+1 - interesting stuff. So this will start a form with a bunch of error messages? Or does it only show them when changes are made?

File: addon/pods/components/rui-validatable-input/component.js [hunk 6]
```diff
@@ -30,25 +33,33 @@ export default Component.extend({
...
   },
 
   // Validation binding attributes with check for async settings
   // if async is true, don't display either until `_didValidate`
 
   isInvalid: computed('_isInvalid', '_didValidate', function() {
     if (this.get('async')) {
-      return this.get('_isInvalid') && this.get('_didValidate')
+      return this.get('_isInvalid') &&
+      this.get('_didValidate')
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: [Style] should be indented.

File: addon/pods/components/rui-validatable-input/component.js [hunk 7]
```diff
@@ -30,25 +33,33 @@ export default Component.extend({
...
     }
     return this.get('_isInvalid')
   }),
 
   isValid: computed('_isValid', '_didValidate', function() {
     if (this.get('async')) {
-      return this.get('_isValid') && this.get('_didValidate')
+      return this.get('_isValid') &&
+      this.get('_didValidate')
```
Comment (kylemellander) [inline addon/pods/components/rui-validatable-input/component.js]: [Style] should be indented.

**Comment by vcolavin** [comment]
@kscherling the setup instructions seem to get confused between Admin and Discuss apps sometimes. For example, there are setup instructions for the discuss app which then direct you to an admin url (`localhost:3000/cli/ea/:project_id/setup/settings/digest/customize`). So I'm not sure where I should be QAing!

**Comment by vcolavin** [comment]
There's an issue with the participant app (url http://localhost:4202/#/login) having jittery validations. I guess the flash error shows up a split second after the button has been pressed, and in the meantime the app displays "Looks good!".

Video here: https://vid.me/BQex

**Comment by vcolavin** [comment]
Sorry for all the noise with these comments! Until I'm added to the org I can't click those QA checkboxes up top so I'll use this list:

### Digest (synchronous validation)
- [ ] Digest inputs show error when empty
- [ ] Show valid when not empty and changed
- [ ] Validation clears after save

### Participant (asynchronousvalidation)
- [ ] Login fields do not display validation when changed
- [ ] When login button clicked, validation shows
- [ ] Once validation is triggered, validators will validate asyncronously

- [ ]  QA complete

**Comment by kscherling** [comment]
@vincent  Looked into the issue.  It looks like this only occurs when you submit the request and it validates, then the server responds with an error.  So it is actually appropriate, though not necessarily great.  I don't see this as a bug though.  I am going to be updating some styles (eventually) that will make this less jarring and with additional response time and loading state on the button should not be quite so 'jittery' looking.
